<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Appendix • SummarizedBenchmark</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Appendix">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SummarizedBenchmark</a>
        <span class="label label-info" data-toggle="tooltip" data-placement="bottom" title="Bioc Devel">1.99.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">General</li>
    <li>
      <a href="../articles/SummarizedBenchmark.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/Appendix.html">Appendix</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Case Studies</li>
    <li>
      <a href="../articles/SingleCellBenchmark.html">scRNA-seq Simulation</a>
    </li>
    <li>
      <a href="../articles/QuantificationBenchmark.html">RNA-seq Quantification</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/areyesq89/SummarizedBenchmark">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Appendix</h1>
                        <h4 class="author">Patrick K. Kimes, Alejandro Reyes</h4>
            
            <h4 class="date">23 October 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/areyesq89/SummarizedBenchmark/blob/master/vignettes/Appendix.Rmd"><code>vignettes/Appendix.Rmd</code></a></small>
      <div class="hidden name"><code>Appendix.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      “”
    </div>
    
<div id="data-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#data-objects" class="anchor"></a>Data objects</h1>
<div id="preparing-the-data-from-soneson-et-al-" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-the-data-from-soneson-et-al-" class="anchor"></a>Preparing the data from Soneson et al.</h2>
<p>The data is available for download from ArrayExpress. Expression data for each sample is provided in the RSEM output format. Corresponding information for the ground truth underlying the simulated data is also available, including transcript differential expression status.</p>
<p>First, we download and import the transcript-level TPM values using the tximport 1.8.0 package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">d &lt;-<span class="st"> </span><span class="kw">tempdir</span>()</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">download.file</span>(<span class="dt">url =</span> <span class="kw">paste0</span>(<span class="st">"https://www.ebi.ac.uk/arrayexpress/files/"</span>,</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">                           <span class="st">"E-MTAB-4119/E-MTAB-4119.processed.3.zip"</span>),</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">              <span class="dt">destfile =</span> <span class="kw">file.path</span>(d, <span class="st">"samples.zip"</span>))</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="kw">unzip</span>(<span class="kw">file.path</span>(d, <span class="st">"samples.zip"</span>), <span class="dt">exdir =</span> d)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">fl &lt;-<span class="st"> </span><span class="kw">list.files</span>(d, <span class="dt">pattern =</span> <span class="st">"*_rsem.txt"</span>, <span class="dt">full.names=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="kw">names</span>(fl) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"sample(.*)_rsem.txt"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw">basename</span>(fl))</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="kw">library</span>(tximport)</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="kw">library</span>(readr)</a>
<a class="sourceLine" id="cb1-11" data-line-number="11"></a>
<a class="sourceLine" id="cb1-12" data-line-number="12">txi &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tximport/topics/tximport">tximport</a></span>(fl, <span class="dt">txIn =</span> <span class="ot">TRUE</span>, <span class="dt">txOut =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb1-13" data-line-number="13">                <span class="dt">geneIdCol =</span> <span class="st">"gene_id"</span>,</a>
<a class="sourceLine" id="cb1-14" data-line-number="14">                <span class="dt">txIdCol =</span> <span class="st">"transcript_id"</span>,</a>
<a class="sourceLine" id="cb1-15" data-line-number="15">                <span class="dt">countsCol =</span> <span class="st">"expected_count"</span>,</a>
<a class="sourceLine" id="cb1-16" data-line-number="16">                <span class="dt">lengthCol =</span> <span class="st">"effective_length"</span>,</a>
<a class="sourceLine" id="cb1-17" data-line-number="17">                <span class="dt">abundanceCol =</span> <span class="st">"TPM"</span>,</a>
<a class="sourceLine" id="cb1-18" data-line-number="18">                <span class="dt">countsFromAbundance =</span> <span class="st">"scaledTPM"</span>,</a>
<a class="sourceLine" id="cb1-19" data-line-number="19">                <span class="dt">importer =</span> <span class="cf">function</span>(x){ <span class="kw"><a href="http://www.rdocumentation.org/packages/readr/topics/read_delim">read_tsv</a></span>(x) })</a></code></pre></div>
<pre><code>## 1 Parsed with column specification:
## cols(
##   transcript_id = col_character(),
##   gene_id = col_character(),
##   length = col_integer(),
##   effective_length = col_double(),
##   expected_count = col_double(),
##   TPM = col_double(),
##   FPKM = col_double(),
##   IsoPct = col_double()
## )
## 2 Parsed with column specification:
## cols(
##   transcript_id = col_character(),
##   gene_id = col_character(),
##   length = col_integer(),
##   effective_length = col_double(),
##   expected_count = col_double(),
##   TPM = col_double(),
##   FPKM = col_double(),
##   IsoPct = col_double()
## )
## 3 Parsed with column specification:
## cols(
##   transcript_id = col_character(),
##   gene_id = col_character(),
##   length = col_integer(),
##   effective_length = col_double(),
##   expected_count = col_double(),
##   TPM = col_double(),
##   FPKM = col_double(),
##   IsoPct = col_double()
## )
## 4 Parsed with column specification:
## cols(
##   transcript_id = col_character(),
##   gene_id = col_character(),
##   length = col_integer(),
##   effective_length = col_double(),
##   expected_count = col_double(),
##   TPM = col_double(),
##   FPKM = col_double(),
##   IsoPct = col_double()
## )
## 5 Parsed with column specification:
## cols(
##   transcript_id = col_character(),
##   gene_id = col_character(),
##   length = col_integer(),
##   effective_length = col_double(),
##   expected_count = col_double(),
##   TPM = col_double(),
##   FPKM = col_double(),
##   IsoPct = col_double()
## )
## 6 Parsed with column specification:
## cols(
##   transcript_id = col_character(),
##   gene_id = col_character(),
##   length = col_integer(),
##   effective_length = col_double(),
##   expected_count = col_double(),
##   TPM = col_double(),
##   FPKM = col_double(),
##   IsoPct = col_double()
## )</code></pre>
<p>Next, we obtain and load the ground truth information that can be used for evaluating the results of the differential expression analysis.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">download.file</span>(<span class="dt">url =</span> <span class="kw">paste0</span>(<span class="st">"https://www.ebi.ac.uk/arrayexpress/files/"</span>,</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">                           <span class="st">"E-MTAB-4119/E-MTAB-4119.processed.2.zip"</span>),</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">              <span class="dt">destfile =</span> <span class="kw">file.path</span>(d, <span class="st">"truth.zip"</span>))</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw">unzip</span>(<span class="kw">file.path</span>(d, <span class="st">"truth.zip"</span>), <span class="dt">exdir =</span> d)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"></a>
<a class="sourceLine" id="cb3-6" data-line-number="6">truthdat &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/readr/topics/read_delim">read_tsv</a></span>(<span class="kw">file.path</span>(d, <span class="st">"truth_transcript.txt"</span>))</a></code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   transcript = col_character(),
##   status = col_integer(),
##   logFC_cat = col_character(),
##   logFC = col_double(),
##   avetpm = col_double(),
##   length = col_integer(),
##   eff_length = col_double(),
##   tpm1 = col_double(),
##   tpm2 = col_double(),
##   isopct1 = col_double(),
##   isopct2 = col_double(),
##   gene = col_character(),
##   avetpm_cat = col_character()
## )</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">#save( txi, truthdat, file="../data/soneson2016.rda" )</span></a></code></pre></div>
</div>
</div>
<div id="non-standard-use" class="section level1">
<h1 class="hasAnchor">
<a href="#non-standard-use" class="anchor"></a>Non-Standard Use</h1>
<div id="manually-constructing-a-summarizedbenchmark" class="section level2">
<h2 class="hasAnchor">
<a href="#manually-constructing-a-summarizedbenchmark" class="anchor"></a>Manually Constructing a SummarizedBenchmark</h2>
<p>So far, this vignette has shown the recommended use of <em>SummarizedBenchmark</em>, that enables users to perform benchmarks automatically keeping track of parameters and software versions. However, users can also construct <em>SummarizedBenchmark</em> objects from standard <code>S3</code> data objects.</p>
<p>Using data from the <em><a href="http://bioconductor.org/packages/iCOBRA">iCOBRA</a></em>package <span class="citation">[@Soneson_2016]</span>, this part of the vignette demonstrates how to build <em>SummarizedBenchmark</em> objects from <code>S3</code> objects. The dataset contains differential expression results of three different methods (<em><a href="http://bioconductor.org/packages/limma">limma</a></em>, <em><a href="http://bioconductor.org/packages/edgeR">edgeR</a></em> and <em><a href="http://bioconductor.org/packages/DESeq2">DESeq2</a></em>) applied to a simulated RNA-seq dataset.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(iCOBRA)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">data</span>(cobradata_example)</a></code></pre></div>
<p>The process of building a <em>SummarizedBenchmark</em> object is similar to the one used to construct a <em>SummarizedExperiment</em> object. To build a <em>SummarizedBenchmark</em> object, three main objects are required (1) a list where each element corresponds to a <em>data.frame</em>, (2) a <em>DataFrame</em> with annotations of the methods and (3) when available, a <em>DataFrame</em> of ground truths.</p>
<p>In the <em>SummarizedBenchmark</em> object, each output of the methods is considered a different <code>assay</code>. For example, using the differential expression dataset example, we can define two assays, q-values and estimated log fold changes. For each <code>assay</code>, we arrange the output of the different methods as a matrix where each column corresponds to a method and each row corresponds to each feature (in this case, genes). We will need a list in which each of it’s element corresponds to an assay.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">assays &lt;-<span class="st"> </span><span class="kw">list</span>(</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">  <span class="dt">qvalue=</span>cobradata_example<span class="op">@</span>padj,</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">  <span class="dt">logFC=</span>cobradata_example<span class="op">@</span>score )</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">assays[[<span class="st">"qvalue"</span>]]<span class="op">$</span>DESeq2 &lt;-<span class="st"> </span><span class="kw">p.adjust</span>(cobradata_example<span class="op">@</span>pval<span class="op">$</span>DESeq2, <span class="dt">method=</span><span class="st">"BH"</span>)</a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="kw">head</span>( assays[[<span class="st">"qvalue"</span>]], <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##                     edgeR      voom     DESeq2
## ENSG00000000457 0.2833295 0.1375146 0.04760152
## ENSG00000000460 0.2700376 0.2826398 0.08717737
## ENSG00000000938 0.9232794 0.9399669 0.99856748</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw">head</span>( assays[[<span class="st">"logFC"</span>]], <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##                      edgeR       voom      DESeq2
## ENSG00000000457 0.22706080 0.23980596 0.274811698
## ENSG00000000460 0.31050478 0.27677537 0.359226203
## ENSG00000000938 0.04985025 0.03344113 0.001526355</code></pre>
<p>Since these are simulated data, the ground truths for both assays are known. We can format these as a <em>DataFrame</em> where each column corresponds to an assay and each row corresponds to a feature.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">library</span>(S4Vectors)</a></code></pre></div>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:iCOBRA':
## 
##     score, score&lt;-</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colMeans, colnames, colSums, dirname, do.call, duplicated,
##     eval, evalq, Filter, Find, get, grep, grepl, intersect,
##     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
##     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
##     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,
##     table, tapply, union, unique, unsplit, which, which.max,
##     which.min</code></pre>
<pre><code>## 
## Attaching package: 'S4Vectors'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     expand.grid</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">groundTruth &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/S4Vectors/topics/DataFrame-class">DataFrame</a></span>( cobradata_example<span class="op">@</span>truth[,<span class="kw"><a href="http://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class">c</a></span>(<span class="st">"status"</span>, <span class="st">"logFC"</span>)] )</a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="kw">colnames</span>(groundTruth) &lt;-<span class="st"> </span><span class="kw">names</span>( assays )</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">groundTruth &lt;-<span class="st"> </span>groundTruth[<span class="kw">rownames</span>(assays[[<span class="dv">1</span>]]),]</a>
<a class="sourceLine" id="cb22-4" data-line-number="4"><span class="kw"><a href="http://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class">head</a></span>( groundTruth )</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##                    qvalue              logFC
##                 &lt;integer&gt;          &lt;numeric&gt;
## ENSG00000000457         0                  0
## ENSG00000000460         0                  0
## ENSG00000000938         0                  0
## ENSG00000000971         0                  0
## ENSG00000001460         1   3.93942204325368
## ENSG00000001461         1 0.0602244207363294</code></pre>
<p>Then, the method names are also reformatted as a <em>DataFrame</em></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">colData &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/S4Vectors/topics/DataFrame-class">DataFrame</a></span>( <span class="dt">method=</span><span class="kw">colnames</span>(assays[[<span class="dv">1</span>]]) )</a>
<a class="sourceLine" id="cb24-2" data-line-number="2">colData</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 1 column
##        method
##   &lt;character&gt;
## 1       edgeR
## 2        voom
## 3      DESeq2</code></pre>
<p>A <em>SummarizedBenchmark</em> is build using the following command</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw">library</span>(SummarizedBenchmark)</a></code></pre></div>
<pre><code>## Loading required package: tidyr</code></pre>
<pre><code>## 
## Attaching package: 'tidyr'</code></pre>
<pre><code>## The following object is masked from 'package:S4Vectors':
## 
##     expand</code></pre>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
<pre><code>## Loading required package: DelayedArray</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: 'matrixStats'</code></pre>
<pre><code>## The following objects are masked from 'package:Biobase':
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## Loading required package: BiocParallel</code></pre>
<pre><code>## 
## Attaching package: 'DelayedArray'</code></pre>
<pre><code>## The following objects are masked from 'package:matrixStats':
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     aperm, apply</code></pre>
<pre><code>## Loading required package: UpSetR</code></pre>
<pre><code>## Loading required package: rlang</code></pre>
<pre><code>## 
## Attaching package: 'rlang'</code></pre>
<pre><code>## The following object is masked from 'package:Biobase':
## 
##     exprs</code></pre>
<pre><code>## Loading required package: stringr</code></pre>
<pre><code>## Loading required package: ggplot2</code></pre>
<pre><code>## Loading required package: mclust</code></pre>
<pre><code>## Package 'mclust' version 5.4.1
## Type 'citation("mclust")' for citing this R package in publications.</code></pre>
<pre><code>## Loading required package: dplyr</code></pre>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following object is masked from 'package:matrixStats':
## 
##     count</code></pre>
<pre><code>## The following object is masked from 'package:Biobase':
## 
##     combine</code></pre>
<pre><code>## The following objects are masked from 'package:GenomicRanges':
## 
##     intersect, setdiff, union</code></pre>
<pre><code>## The following object is masked from 'package:GenomeInfoDb':
## 
##     intersect</code></pre>
<pre><code>## The following objects are masked from 'package:IRanges':
## 
##     collapse, desc, intersect, setdiff, slice, union</code></pre>
<pre><code>## The following objects are masked from 'package:S4Vectors':
## 
##     first, intersect, rename, setdiff, setequal, union</code></pre>
<pre><code>## The following objects are masked from 'package:BiocGenerics':
## 
##     combine, intersect, setdiff, union</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre><code>## Loading required package: digest</code></pre>
<pre><code>## Loading required package: sessioninfo</code></pre>
<pre><code>## Loading required package: crayon</code></pre>
<pre><code>## 
## Attaching package: 'crayon'</code></pre>
<pre><code>## The following object is masked from 'package:ggplot2':
## 
##     %+%</code></pre>
<pre><code>## The following object is masked from 'package:rlang':
## 
##     chr</code></pre>
<pre><code>## Loading required package: tibble</code></pre>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb70-1" data-line-number="1">sb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SummarizedBenchmark.html">SummarizedBenchmark</a></span>(</a>
<a class="sourceLine" id="cb70-2" data-line-number="2">  <span class="dt">assays=</span>assays, </a>
<a class="sourceLine" id="cb70-3" data-line-number="3">  <span class="dt">colData=</span>colData,</a>
<a class="sourceLine" id="cb70-4" data-line-number="4">  <span class="dt">groundTruth=</span>groundTruth )</a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#data-objects">Data objects</a><ul class="nav nav-pills nav-stacked">
<li><a href="#preparing-the-data-from-soneson-et-al-">Preparing the data from Soneson et al.</a></li>
      </ul>
</li>
      <li>
<a href="#non-standard-use">Non-Standard Use</a><ul class="nav nav-pills nav-stacked">
<li><a href="#manually-constructing-a-summarizedbenchmark">Manually Constructing a SummarizedBenchmark</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://alejandroreyes.org">Alejandro Reyes</a>, <a href="https://www.pkimes.com">Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
