<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SummarizedBenchmark: Class Details • SummarizedBenchmark</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SummarizedBenchmark: Class Details">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SummarizedBenchmark</a>
        <span class="version label label-info" data-toggle="tooltip" data-placement="bottom" title="Bioc Devel">2.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/SummarizedBenchmark-Introduction.html">Quick Start</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">General</li>
    <li>
      <a href="../articles/SummarizedBenchmark-Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/SummarizedBenchmark-ClassDetails.html">Class Details</a>
    </li>
    <li>
      <a href="../articles/SummarizedBenchmark-FullCaseStudy.html">Full Case Study</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Features</li>
    <li>
      <a href="../articles/Feature-ErrorHandling.html">Error Handling</a>
    </li>
    <li>
      <a href="../articles/Feature-Iterative.html">Iterative Benchmarking</a>
    </li>
    <li>
      <a href="../articles/Feature-Parallel.html">Parallelization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">More Case Studies</li>
    <li>
      <a href="../articles/CaseStudy-SingleCellSimulation.html">scRNA-seq Simulation</a>
    </li>
    <li>
      <a href="../articles/CaseStudy-RNAseqQuantification.html">RNA-seq Quantification</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/areyesq89/SummarizedBenchmark">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>SummarizedBenchmark: Class Details</h1>
                        <h4 class="author">Patrick K. Kimes, Alejandro Reyes</h4>
            
            <h4 class="date">19 June 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/areyesq89/SummarizedBenchmark/blob/master/vignettes/SummarizedBenchmark-ClassDetails.Rmd"><code>vignettes/SummarizedBenchmark-ClassDetails.Rmd</code></a></small>
      <div class="hidden name"><code>SummarizedBenchmark-ClassDetails.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      “The <code>SummarizedBenchmark</code> package is built on two primary classes, the <em>BenchDesign</em> class for organizing the design of a benchmark experiment, and the <em>SummarizedExperiment</em> class for organizing the benchmark results. This vignette first introduces the structure and features of these classes using the example benchmark of p-value correction methods described in the <strong>SummarizedBenchark: Introduction</strong> vignette. Then, some issues of reproduciblity are discussed, and finally, several features and non-standard uses of the classes are described. SummarizedBenchmark package version: 2.3.4”
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette assumes that the user is familiar with the basics of the <code>SummarizedBenchmark</code> framework for benchmarking computational methods described in the <strong>SummarizedBenchmark: Introduction</strong> vignette (the “introduction”). New users should start from the package description and example presented in the introduction before moving on to more advanced topics, including the class descriptions in this vignette. Here, we describe the anatomy of the <em>BenchDesign</em> and <em>SummarizedBenchmark</em> classes in a little more detail than was presented in the introduction.</p>
<div id="example-case-study" class="section level2">
<h2 class="hasAnchor">
<a href="#example-case-study" class="anchor"></a>Example Case Study</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"SummarizedBenchmark"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"magrittr"</span>)</a></code></pre></div>
<p>The classes are described in more detail using the same example benchmark of methods for multiple comparisons adjustment described in the introduction. Briefly, three methods, the Bonferroni correction, Benjamini-Hochberg step-up procedure, and Storey’s FDR q-value were benchmarked on a collection of 50 simulated two-sample t-tests. The simulated t-test results which serve as inputs to the methods are included with the package in the <code>tdat</code> data set.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(tdat)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"></a>
<a class="sourceLine" id="cb2-3" data-line-number="3">b &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>(<span class="dt">data =</span> tdat) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"bonf"</span>, <span class="dt">func =</span> p.adjust,</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval, <span class="dt">method =</span> <span class="st">"bonferroni"</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"BH"</span>,</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">              <span class="dt">func =</span> p.adjust,</a>
<a class="sourceLine" id="cb2-8" data-line-number="8">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval, <span class="dt">method =</span> <span class="st">"BH"</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"qv"</span>,</a>
<a class="sourceLine" id="cb2-10" data-line-number="10">              <span class="dt">func =</span> qvalue<span class="op">::</span>qvalue,</a>
<a class="sourceLine" id="cb2-11" data-line-number="11">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval),</a>
<a class="sourceLine" id="cb2-12" data-line-number="12">              <span class="dt">post =</span> <span class="cf">function</span>(x) { x<span class="op">$</span>qvalues })</a>
<a class="sourceLine" id="cb2-13" data-line-number="13"></a>
<a class="sourceLine" id="cb2-14" data-line-number="14">sb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(b, <span class="dt">truthCols =</span> <span class="st">"H"</span>)</a></code></pre></div>
<p>The code above is identical to what was used in the introduction to construct the <em>BenchDesign</em>, <code>b</code>, and the <em>SummarizedBenchmark</em> results, <code>sb</code>. If any of this is unclear, consider revisting the <strong>SummarizedBenchmark: Introduction</strong> vignette.</p>
</div>
</div>
<div id="benchdesign" class="section level1">
<h1 class="hasAnchor">
<a href="#benchdesign" class="anchor"></a>BenchDesign</h1>
<p><em>BenchDesign</em> objects are composed of a data set and methods. Formally, the methods are stored in the <em>BenchDesign</em> as a <em>BDMethodList</em> and the data as a <em>BDData</em> object. As would be expected, the <em>BDMethodList</em> is a list (<code>List</code>) of <em>BDMethod</em> objects, each containing the definition of a method to be compared in the benchmark experiment. The general structure is illustrated in the figure below.</p>
<div class="figure">
<img src="summarizedbenchmark-figure2.png" alt="BenchDesign class structure"><p class="caption">BenchDesign class structure</p>
</div>
<p>Continuing with the <em>BenchDesign</em> constructed above, we can access the list of methods and the data set stored in the object using the <code><a href="../reference/BDMethodList.html">BDMethodList()</a></code> and <code><a href="../reference/BDData.html">BDData()</a></code> functions.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="../reference/BDMethodList.html">BDMethodList</a></span>(b)</a></code></pre></div>
<pre><code>## BenchDesign Method List (BDMethodList) --------------------- 
## list of 3 methods 
##   method:  bonf; func: p.adjust       
##   method:    BH; func: p.adjust       
##   method:    qv; func: qvalue::qvalue</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="../reference/BDData.html">BDData</a></span>(b)</a></code></pre></div>
<pre><code>## BenchDesign Data (BDData) ---------------------------------- 
##   type: data 
##   data:
##    H test_statistic effect_size         pval        SE
## 1  1     -3.2083437 -1.17151466 4.872282e-03 0.3651463
## 2  0      0.1692236  0.07321912 8.675080e-01 0.4326768
## 3  1     -5.7077940 -1.81715381 2.061521e-05 0.3183636
## 4  1     -1.9805856 -1.09107836 6.313031e-02 0.5508867
## 5  1     -1.0014358 -0.37726058 3.298895e-01 0.3767197
## 6  1     -0.9190433 -0.47583669 3.702252e-01 0.5177522
## 7  1     -1.3184486 -0.72750221 2.038846e-01 0.5517866
## 8  1     -3.8492346 -1.47787871 1.175116e-03 0.3839409
## 9  1     -3.4836213 -1.35474550 2.651401e-03 0.3888900
## 10 0      0.6238195  0.25470364 5.405719e-01 0.4082970
##  ... with 30 more rows.</code></pre>
<div id="bdmethod-and-bdmethodlist" class="section level2">
<h2 class="hasAnchor">
<a href="#bdmethod-and-bdmethodlist" class="anchor"></a>BDMethod and BDMethodList</h2>
<p>The list of methods inherits from the <em>SimpleList</em> class, and supports several useful accessor and setter features.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="../reference/BDMethodList.html">BDMethodList</a></span>(b)[[<span class="st">"bonf"</span>]]</a></code></pre></div>
<pre><code>## BenchDesign Method (BDMethod) ------------------------------ 
##   method: p.adjust
##     function (p, method = p.adjust.methods, n = length(p)... 
##   parameters:
##     p : pval 
##     method : "bonferroni" 
##   post:
##     none
##   meta:
##     none</code></pre>
<p>This interface allows for adding new methods by creating a new named entry in the <em>BDMethodList</em> of the <em>BenchDesign</em> object.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="../reference/BDMethodList.html">BDMethodList</a></span>(b)[[<span class="st">"bonf2"</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BDMethodList.html">BDMethodList</a></span>(b)[[<span class="st">"bonf"</span>]]</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">b</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     type: data 
##     names: H, test_statistic, effect_size, pval, SE 
##   benchmark methods:
##     method:   bonf; func: p.adjust       
##     method:     BH; func: p.adjust       
##     method:     qv; func: qvalue::qvalue 
##     method:  bonf2; func: p.adjust</code></pre>
<p>Methods can also be removed by setting the list value to NULL. An equivalent convenience function, <code><a href="../reference/dropMethod.html">dropMethod()</a></code> also exists for removing methods using syntax similar to <code><a href="../reference/addMethod.html">addMethod()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="../reference/BDMethodList.html">BDMethodList</a></span>(b)[[<span class="st">"bonf"</span>]] &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2">b</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     type: data 
##     names: H, test_statistic, effect_size, pval, SE 
##   benchmark methods:
##     method:     BH; func: p.adjust       
##     method:     qv; func: qvalue::qvalue 
##     method:  bonf2; func: p.adjust</code></pre>
<p>Each <em>BDMethod</em> object in a <em>BDMethodList</em> encapsulates a method to be benchmarked. The contents of the object are all of the arguments passed through the <code><a href="../reference/addMethod.html">addMethod()</a></code> convenience function used in the example above. New <em>BDMethod</em> objects can be constructed directly using the <code><a href="../reference/BDMethod.html">BDMethod()</a></code> constructor, which takes the same input parameters as <code><a href="../reference/addMethod.html">addMethod()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">bdm_bonf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BDMethod.html">BDMethod</a></span>(<span class="dt">x =</span> p.adjust, </a>
<a class="sourceLine" id="cb13-2" data-line-number="2">                     <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval, </a>
<a class="sourceLine" id="cb13-3" data-line-number="3">                                          <span class="dt">method =</span> <span class="st">"bonferroni"</span>))</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">bdm_bonf</a></code></pre></div>
<pre><code>## BenchDesign Method (BDMethod) ------------------------------ 
##   method: p.adjust
##     function (p, method = p.adjust.methods, n = length(p)... 
##   parameters:
##     p : pval 
##     method : "bonferroni" 
##   post:
##     none
##   meta:
##     none</code></pre>
<p>Directly modifying the <em>BDMethodList</em> object provides an alternative approach (aside form using <code><a href="../reference/addMethod.html">addMethod()</a></code>) to adding methods to a <em>BenchDesign</em> object.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="../reference/BDMethodList.html">BDMethodList</a></span>(b)[[<span class="st">"bonf"</span>]] &lt;-<span class="st"> </span>bdm_bonf</a>
<a class="sourceLine" id="cb15-2" data-line-number="2">b</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     type: data 
##     names: H, test_statistic, effect_size, pval, SE 
##   benchmark methods:
##     method:     BH; func: p.adjust       
##     method:     qv; func: qvalue::qvalue 
##     method:  bonf2; func: p.adjust       
##     method:   bonf; func: p.adjust</code></pre>
</div>
<div id="bddata" class="section level2">
<h2 class="hasAnchor">
<a href="#bddata" class="anchor"></a>BDData</h2>
<p>The <em>BDData</em> object is a simple object which only contains two slots, <code>@type</code> and <code>@data</code>. The <code>@type</code> of a <em>BDData</em> object can be either <code>data</code> or <code>hash</code>. If the <code>@type</code> is <code>data</code>, the <code>@data</code> slot is a standard list or data.frame object of raw data. However, if the <code>@type</code> is <code>hash</code>, then the <code>@data</code> slot is just a MD5 hash of some raw data object computed using <code><a href="https://www.rdocumentation.org/packages/digest/topics/digest">digest::digest()</a></code>. As described next in Section @ref(summarizedbenchmark), <em>SummarizedBenchmark</em> objects also contain the <em>BenchDesign</em> object used to generate the benchmark results. Often, the original raw data can be large, and saving the full data set as part of the <em>SummarizedBenchmark</em> object can be undesirable and unnecessary. While the raw data set is needed during the benchmark experiment, by default, the <em>BDData</em> is converted to a MD5 hash beforing stored the <em>BenchDesign</em> as part of a <em>SummarizedBenchmark</em> object. Using this approach, it is still possible to confirm whether a certain data set was used for a benchmarking experiment without having to store copies of the raw data with every <em>SummarizedBenchmark</em>.</p>
<p>Returning to the <em>BenchDesign</em> example from above, the <em>BDData</em> object of a <em>BenchDesign</em> can also be similarly extracted and modified. As with <em>BDMethod</em> objects, the data set may be removed by setting the value to NULL.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="../reference/BDData.html">BDData</a></span>(b) &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">b</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     NULL
##   benchmark methods:
##     method:     BH; func: p.adjust       
##     method:     qv; func: qvalue::qvalue 
##     method:  bonf2; func: p.adjust       
##     method:   bonf; func: p.adjust</code></pre>
<p>A new data set can be constructed using the <code><a href="../reference/BDData.html">BDData()</a></code> function and used to replace the data set in a <em>BenchDesign</em> object.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">bdd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BDData.html">BDData</a></span>(<span class="dt">data =</span> tdat)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw"><a href="../reference/BDData.html">BDData</a></span>(b) &lt;-<span class="st"> </span>bdd</a>
<a class="sourceLine" id="cb19-3" data-line-number="3">b</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     type: data 
##     names: H, test_statistic, effect_size, pval, SE 
##   benchmark methods:
##     method:     BH; func: p.adjust       
##     method:     qv; func: qvalue::qvalue 
##     method:  bonf2; func: p.adjust       
##     method:   bonf; func: p.adjust</code></pre>
<p>The helper function <code><a href="../reference/hashBDData.html">hashBDData()</a></code> can be called on either <em>BDData</em> or <em>BenchDesign</em> objects to convert the data to a MD5 hash. By default, this function is called on the <em>BenchDesign</em> object during <code><a href="../reference/buildBench.html">buildBench()</a></code> before storing the object in the resulting <em>SummarizedBenchmark</em>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">b &lt;-<span class="st"> </span><span class="kw"><a href="../reference/hashBDData.html">hashBDData</a></span>(b)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">b</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     type: md5hash 
##     MD5 hash: 09867f2696d3178a8f3bc038917042e1 
##   benchmark methods:
##     method:     BH; func: p.adjust       
##     method:     qv; func: qvalue::qvalue 
##     method:  bonf2; func: p.adjust       
##     method:   bonf; func: p.adjust</code></pre>
<p>We drop the <code>"bonf2"</code> method from the object to return to the same set of methods used to generate the <em>SummarizedBenchmark</em> object.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw"><a href="../reference/BDMethodList.html">BDMethodList</a></span>(b)[[<span class="st">"bonf2"</span>]] &lt;-<span class="st"> </span><span class="ot">NULL</span></a></code></pre></div>
</div>
</div>
<div id="summarizedbenchmark" class="section level1">
<h1 class="hasAnchor">
<a href="#summarizedbenchmark" class="anchor"></a>SummarizedBenchmark</h1>
<p>As described above, the <em>SummarizedBenchmark</em> class builds on the existing <em>SummarizedExperiment</em> class and Bioconductor infrastructure. By doing so, saved results are tightly linked to metadata. Thus, it is possible, for example, to easily subset the results without losing references to the respective metadata. For example, the code below extracts the data for only the first two methods.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">sbSub &lt;-<span class="st"> </span>sb[, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</a>
<a class="sourceLine" id="cb24-2" data-line-number="2"><span class="kw">colData</span>(sbSub)</a></code></pre></div>
<pre><code>## DataFrame with 2 rows and 6 columns
##         func.pkg func.pkg.vers func.pkg.manual     param.p param.method
##      &lt;character&gt;   &lt;character&gt;       &lt;logical&gt; &lt;character&gt;  &lt;character&gt;
## bonf       stats         3.6.0           FALSE        pval "bonferroni"
## BH         stats         3.6.0           FALSE        pval         "BH"
##      session.idx
##        &lt;numeric&gt;
## bonf           1
## BH             1</code></pre>
<p>Building on the <em>SummarizedExperiment</em> class, in addition to the slot for <code>performanceMetrics</code> described above, the <em>SummarizedBenchmark</em> class also includes a slot which stores a copy of the <em>BenchDesign</em> object used to generate the results with <code><a href="../reference/buildBench.html">buildBench()</a></code>. The <em>BenchDesign</em> can be accessed by simply passing the object to <code><a href="../reference/BenchDesign.html">BenchDesign()</a></code>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>(sb)</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     type: md5hash 
##     MD5 hash: 09867f2696d3178a8f3bc038917042e1 
##   benchmark methods:
##     method:  bonf; func: p.adjust       
##     method:    BH; func: p.adjust       
##     method:    qv; func: qvalue::qvalue</code></pre>
<p>Notice, however, unlike the original <em>BenchDesign</em> above, the data is stored as a MD5 hash. To prevent bloat when the data set is large, by default, only a MD5 hash computed using <code><a href="https://www.rdocumentation.org/packages/digest/topics/digest">digest::digest()</a></code> is stored with the benchmarking results. This behavior can be changed by setting <code>keepData = TRUE</code> when calling <code><a href="../reference/buildBench.html">buildBench()</a></code>, in which case, the complete data set will be stored as part of the <em>BenchDesign</em> and <em>SummarizedBenchmark</em> objects. A simple helper function, <code><a href="../reference/compareBDData.html">compareBDData()</a></code>, can be used to verify that while the data <em>type</em> may be different between the data originally used to construct the <em>SummarizedBenchmark</em> and the data stored with the <em>SummarizedBenchmark</em>, the referenced data sets are the same (i.e. have matching MD5 hash values).</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw"><a href="../reference/compareBDData.html">compareBDData</a></span>(<span class="kw"><a href="../reference/BDData.html">BDData</a></span>(tdat), <span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>(sb))</a></code></pre></div>
<pre><code>## $data
## [1] TRUE
## 
## $type
## [1] FALSE</code></pre>
<p>Finally, information about the R session where the methods were executed is stored in the metadata of the <em>SummarizedBenchmark</em> as an entry in a list called <code>sessions</code>. The package supports rerunning benchmark experiments, and for this purpose, each new session is stored as a new entry in the <code>sessions</code> list. Since this experiment has only been executed once, the list only has a single entry in the <code>sessions</code> list. In addition to storing the session info and list of methods evaluated during the session, the list also keeps track of all parameters in the <code><a href="../reference/buildBench.html">buildBench()</a></code> call, i.e. <code>truthCols = "H"</code>.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="kw">metadata</span>(sb)<span class="op">$</span>sessions[[<span class="dv">1</span>]]</a></code></pre></div>
<pre><code>## $methods
## [1] "bonf" "BH"   "qv"  
## 
## $results
## $results$bonf
## $results$bonf$default
## [1] "success"
## 
## 
## $results$BH
## $results$BH$default
## [1] "success"
## 
## 
## $results$qv
## $results$qv$default
## [1] "success"
## 
## 
## 
## $parameters
## $parameters$truthCols
## [1] "H"
## 
## $parameters$ftCols
## NULL
## 
## $parameters$sortIDs
## [1] FALSE
## 
## $parameters$keepData
## [1] FALSE
## 
## $parameters$catchErrors
## [1] TRUE
## 
## $parameters$parallel
## [1] FALSE
## 
## $parameters$BPPARAM
## class: MulticoreParam
##   bpisup: FALSE; bpnworkers: 2; bptasks: 0; bpjobname: BPJOB
##   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE
##   bpRNGseed: ; bptimeout: 2592000; bpprogressbar: FALSE
##   bpexportglobals: TRUE
##   bplogdir: NA
##   bpresultdir: NA
##   cluster type: FORK
## 
## 
## $sessionInfo
## ─ Session info ──────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 3.6.0 (2019-04-26)
##  os       macOS Mojave 10.14.1        
##  system   x86_64, darwin15.6.0        
##  ui       X11                         
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  ctype    en_US.UTF-8                 
##  tz       America/New_York            
##  date     2019-06-18                  
## 
## ─ Packages ──────────────────────────────────────────────────────────────
##  package              * version   date       lib source        
##  assertthat             0.2.1     2019-03-21 [1] CRAN (R 3.6.0)
##  backports              1.1.4     2019-04-10 [1] CRAN (R 3.6.0)
##  Biobase              * 2.45.0    2019-05-02 [1] Bioconductor  
##  BiocGenerics         * 0.31.4    2019-06-10 [1] Bioconductor  
##  BiocManager            1.30.4    2018-11-13 [1] CRAN (R 3.6.0)
##  BiocParallel         * 1.19.0    2019-05-02 [1] Bioconductor  
##  BiocStyle            * 2.13.2    2019-06-12 [1] Bioconductor  
##  bitops                 1.0-6     2013-08-17 [1] CRAN (R 3.6.0)
##  bookdown               0.11      2019-05-28 [1] CRAN (R 3.6.0)
##  cli                    1.1.0     2019-03-19 [1] CRAN (R 3.6.0)
##  codetools              0.2-16    2018-12-24 [2] CRAN (R 3.6.0)
##  colorspace             1.4-1     2019-03-18 [1] CRAN (R 3.6.0)
##  commonmark             1.7       2018-12-01 [1] CRAN (R 3.6.0)
##  crayon               * 1.3.4     2017-09-16 [1] CRAN (R 3.6.0)
##  DelayedArray         * 0.11.1    2019-06-16 [1] Bioconductor  
##  desc                   1.2.0     2018-05-01 [1] CRAN (R 3.6.0)
##  digest               * 0.6.19    2019-05-20 [1] CRAN (R 3.6.0)
##  dplyr                * 0.8.1     2019-05-14 [1] CRAN (R 3.6.0)
##  evaluate               0.14      2019-05-28 [1] CRAN (R 3.6.0)
##  fs                     1.3.1     2019-05-06 [1] CRAN (R 3.6.0)
##  GenomeInfoDb         * 1.21.1    2019-05-16 [1] Bioconductor  
##  GenomeInfoDbData       1.2.1     2019-06-17 [1] Bioconductor  
##  GenomicRanges        * 1.37.11   2019-06-11 [1] Bioconductor  
##  ggplot2              * 3.2.0     2019-06-16 [1] CRAN (R 3.6.0)
##  glue                   1.3.1     2019-03-12 [1] CRAN (R 3.6.0)
##  gridExtra              2.3       2017-09-09 [1] CRAN (R 3.6.0)
##  gtable                 0.3.0     2019-03-25 [1] CRAN (R 3.6.0)
##  htmltools              0.3.6     2017-04-28 [1] CRAN (R 3.6.0)
##  IRanges              * 2.19.10   2019-06-11 [1] Bioconductor  
##  knitr                  1.23      2019-05-18 [1] CRAN (R 3.6.0)
##  lattice                0.20-38   2018-11-04 [2] CRAN (R 3.6.0)
##  lazyeval               0.2.2     2019-03-15 [1] CRAN (R 3.6.0)
##  magrittr             * 1.5       2014-11-22 [1] CRAN (R 3.6.0)
##  MASS                   7.3-51.4  2019-03-31 [2] CRAN (R 3.6.0)
##  Matrix                 1.2-17    2019-03-22 [2] CRAN (R 3.6.0)
##  matrixStats          * 0.54.0    2018-07-23 [1] CRAN (R 3.6.0)
##  mclust               * 5.4.3     2019-03-14 [1] CRAN (R 3.6.0)
##  memoise                1.1.0     2017-04-21 [1] CRAN (R 3.6.0)
##  munsell                0.5.0     2018-06-12 [1] CRAN (R 3.6.0)
##  pillar                 1.4.1     2019-05-28 [1] CRAN (R 3.6.0)
##  pkgconfig              2.0.2     2018-08-16 [1] CRAN (R 3.6.0)
##  pkgdown                1.3.0     2018-12-07 [1] CRAN (R 3.6.0)
##  plyr                   1.8.4     2016-06-08 [1] CRAN (R 3.6.0)
##  purrr                  0.3.2     2019-03-15 [1] CRAN (R 3.6.0)
##  qvalue                 2.17.0    2019-05-02 [1] Bioconductor  
##  R6                     2.4.0     2019-02-14 [1] CRAN (R 3.6.0)
##  Rcpp                   1.0.1     2019-03-17 [1] CRAN (R 3.6.0)
##  RCurl                  1.95-4.12 2019-03-04 [1] CRAN (R 3.6.0)
##  reshape2               1.4.3     2017-12-11 [1] CRAN (R 3.6.0)
##  rlang                * 0.3.4     2019-04-07 [1] CRAN (R 3.6.0)
##  rmarkdown              1.13      2019-05-22 [1] CRAN (R 3.6.0)
##  roxygen2               6.1.1     2018-11-07 [1] CRAN (R 3.6.0)
##  rprojroot              1.3-2     2018-01-03 [1] CRAN (R 3.6.0)
##  S4Vectors            * 0.23.12   2019-06-11 [1] Bioconductor  
##  scales                 1.0.0     2018-08-09 [1] CRAN (R 3.6.0)
##  sessioninfo          * 1.1.1     2018-11-05 [1] CRAN (R 3.6.0)
##  stringi                1.4.3     2019-03-12 [1] CRAN (R 3.6.0)
##  stringr              * 1.4.0     2019-02-10 [1] CRAN (R 3.6.0)
##  SummarizedBenchmark  * 2.3.4     2019-06-18 [1] Bioconductor  
##  SummarizedExperiment * 1.15.3    2019-06-16 [1] Bioconductor  
##  tibble               * 2.1.3     2019-06-06 [1] CRAN (R 3.6.0)
##  tidyr                * 0.8.3     2019-03-01 [1] CRAN (R 3.6.0)
##  tidyselect             0.2.5     2018-10-11 [1] CRAN (R 3.6.0)
##  UpSetR               * 1.4.0     2019-05-22 [1] CRAN (R 3.6.0)
##  withr                  2.1.2     2018-03-15 [1] CRAN (R 3.6.0)
##  xfun                   0.7       2019-05-14 [1] CRAN (R 3.6.0)
##  xml2                   1.2.0     2018-01-24 [1] CRAN (R 3.6.0)
##  XVector                0.25.0    2019-05-02 [1] Bioconductor  
##  yaml                   2.2.0     2018-07-25 [1] CRAN (R 3.6.0)
##  zlibbioc               1.31.0    2019-05-02 [1] Bioconductor  
## 
## [1] /usr/local/RLibDevel
## [2] /Library/Frameworks/R.framework/Versions/3.6/Resources/library</code></pre>
</div>
<div id="reproducibility-and-replicability" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility-and-replicability" class="anchor"></a>Reproducibility and Replicability</h1>
<p>While the <em>BenchDesign</em> and <em>SummarizedBenchmark</em> classes are designed to enable easier reproduction and replication of benchmark experiments, they are not meant to completely replace properly written and documented analyses. If any functions or packages are needed for the methods defined in the <em>BenchDesign</em>, they must be available and sourced or loaded when running (or re-running) the analyses. As an example, suppose we have the following locally defined method, <code>mymethod()</code>, which depends on another locally defined, but unavailable, function, <code>mysubmethod()</code>.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">mymethod &lt;-<span class="st"> </span><span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb32-2" data-line-number="2">    <span class="kw">mysubmethod</span>(x)</a>
<a class="sourceLine" id="cb32-3" data-line-number="3">}</a>
<a class="sourceLine" id="cb32-4" data-line-number="4">mybd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">vals =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>))</a>
<a class="sourceLine" id="cb32-5" data-line-number="5">mybd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(mybd, <span class="st">"method1"</span>, mymethod,</a>
<a class="sourceLine" id="cb32-6" data-line-number="6">                  <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">x =</span> vals))</a></code></pre></div>
<p>The top level <code>mymethod()</code> is captured in the corresponding <em>BDMethod</em> and <em>BenchDesign</em>. No errors are thrown since including <code>mymethod()</code> in a <em>BenchDesign</em> object does not require evaluating the function.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw"><a href="../reference/BDMethod.html">BDMethod</a></span>(mybd, <span class="st">"method1"</span>)<span class="op">@</span>f</a></code></pre></div>
<pre><code>## function(x) {
##     mysubmethod(x)
## }</code></pre>
<p>However, when <code><a href="../reference/buildBench.html">buildBench()</a></code> is called, the method naturally fails because the necessary <code>mysubmethod()</code> is not available in the current session.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/conditions">tryCatch</a></span>({ <span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(mybd) }, <span class="dt">error =</span> <span class="cf">function</span>(e) <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(e))</a></code></pre></div>
<pre><code>## !! error caught in buildBench !!
## !! error in main function of method: 'method1'</code></pre>
<pre><code>## !!  original message: 
## !!  could not find function "mysubmethod"</code></pre>
<pre><code>## &lt;simpleError in buildBench(mybd): 
## No method returned valid values for any assays!
## &gt;</code></pre>
<p>Therefore, if any locally defined functions are necessary for a method, they should be available along with the <em>BenchDesign</em> object or defined at the top level. Additionally, for reproducibility and clarity, we generally recommend keeping functions as “self-contained” as possible, and <em>never</em> relying on variables defined externally, e.g. in the global enviroment. Scoping with functions can be confusing, and it may not always be clear where constants are being evaluated. As an example, consider the following simple multiplying function.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1">m &lt;-<span class="st"> </span><span class="dv">5</span></a>
<a class="sourceLine" id="cb39-2" data-line-number="2">mymult &lt;-<span class="st"> </span><span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb39-3" data-line-number="3">    m <span class="op">*</span><span class="st"> </span>x</a>
<a class="sourceLine" id="cb39-4" data-line-number="4">}</a>
<a class="sourceLine" id="cb39-5" data-line-number="5">m &lt;-<span class="st"> </span><span class="dv">2</span></a>
<a class="sourceLine" id="cb39-6" data-line-number="6"></a>
<a class="sourceLine" id="cb39-7" data-line-number="7">mybd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">vals =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">m =</span> <span class="dv">10</span>))</a>
<a class="sourceLine" id="cb39-8" data-line-number="8">mybd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(mybd, <span class="st">"methodr"</span>, mymult,</a>
<a class="sourceLine" id="cb39-9" data-line-number="9">                  <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">x =</span> vals))</a></code></pre></div>
<p>While experienced R programmers may know which value of <code>m</code> (5, 2 or 10) will be used when <code><a href="../reference/buildBench.html">buildBench()</a></code> is called, for many users, this is less obvious.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="kw">assay</span>(<span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(mybd))</a></code></pre></div>
<pre><code>##      methodr
## [1,]       2
## [2,]       4
## [3,]       6
## [4,]       8
## [5,]      10</code></pre>
<p>The answer is the value assigned most recently in the environment where <code>mymult()</code> was defined, <code>R_GlobalEnv</code>, the global environment (<code>m = 2</code>). Note, however, that if the <em>BenchDesign</em> is saved and reloaded in a new session, whatever value of <code>m</code> is defined in the global environment of the new session will be used. If <code>m</code> is not defined in the new session, an error will be thrown. In this case, <code>m</code> should either be explicitly defined within the function or passed as a second input variable of <code>mymult()</code> and defined with <code>params =</code> in the <em>BDMethod</em>.</p>
<p>While it may be possible to aggressively capture all code and environment variables defined with each method in a <em>BenchDesign</em>, it is not within the scope of this package, and furthermore, may not be the best appraoch to constructing a reproducible benchmark. While redundant, for emphasis, we reiterate that the <code>SummarizedBenchmark</code> framework is a solution for structuring benchmark experiments which complements, but does not replace, well documented and organized code.</p>
</div>
<div id="benchdesign-features" class="section level1">
<h1 class="hasAnchor">
<a href="#benchdesign-features" class="anchor"></a>BenchDesign Features</h1>
<p>Several less prominent, but hopefully useful features are also implemented for working with objects of the <em>BenchDesign</em>, <em>BDMethod</em>, and <em>BDMethodList</em> classes.</p>
<div id="specifying-method-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#specifying-method-metadata" class="anchor"></a>Specifying Method Metadata</h2>
<p>As seen in the example above, metadata for methods are stored in the <code>colData()</code> of <em>SummarizedBenchmark</em> objects. Several default metadata columns are populated in the <code>colData()</code> of the <em>SummarizedBenchmark</em> object generated by a call to <code><a href="../reference/buildBench.html">buildBench()</a></code>. However, sometimes it may be useful to include additional metadata columns beyond just the defaults. While this can be accomplished manually by modifying the <code>colData()</code> of the <em>SummarizedBenchmark</em> object post hoc, method metadata can also be specified at the <code><a href="../reference/addMethod.html">addMethod()</a></code> step using the <code>meta =</code> optional parameter. The <code>meta =</code> parameter accepts a <strong>named list</strong> of metadata information. Each list entry will be added to the <code>colData()</code> as a new column, and should be the same type across methods to prevent possible errors. To avoid collisions between metadata columns specified with <code>meta =</code> and the default set of columns, metadata specified using <code>meta =</code> will be added to <code>colData()</code> with <code>meta.</code> prefixed to the column name.</p>
<p>As an example, we construct a <em>BenchDesign</em> object again using the multiple hypothesis testing example. The <em>BenchDesign</em> is created with two methods, Bonferroni correction and the Benjamini-Hochberg step-up procedure. Each method is specified with the optional <code>meta =</code> parameter noting the type of control provided, either familywise error rate (FWER) or false discovery rate (FDR). We can verify that the manually defined metadata column (<code>meta.type</code>) is available in the <code>colData()</code> of the newly generated <em>SummarizedBenchmark</em>.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1">b_withmeta &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>(<span class="dt">data =</span> tdat) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"bonf"</span>, <span class="dt">func =</span> p.adjust,</a>
<a class="sourceLine" id="cb42-3" data-line-number="3">              <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">type =</span> <span class="st">"FWER control"</span>),</a>
<a class="sourceLine" id="cb42-4" data-line-number="4">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval, <span class="dt">method =</span> <span class="st">"bonferroni"</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb42-5" data-line-number="5"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"BH"</span>,</a>
<a class="sourceLine" id="cb42-6" data-line-number="6">              <span class="dt">func =</span> p.adjust,</a>
<a class="sourceLine" id="cb42-7" data-line-number="7">              <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">type =</span> <span class="st">"FDR control"</span>),</a>
<a class="sourceLine" id="cb42-8" data-line-number="8">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval, <span class="dt">method =</span> <span class="st">"BH"</span>))</a>
<a class="sourceLine" id="cb42-9" data-line-number="9"></a>
<a class="sourceLine" id="cb42-10" data-line-number="10">sb_withmeta &lt;-<span class="st"> </span><span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(b_withmeta)</a>
<a class="sourceLine" id="cb42-11" data-line-number="11"><span class="kw">colData</span>(sb_withmeta)</a></code></pre></div>
<pre><code>## DataFrame with 2 rows and 7 columns
##         func.pkg func.pkg.vers func.pkg.manual     param.p param.method
##      &lt;character&gt;   &lt;character&gt;       &lt;logical&gt; &lt;character&gt;  &lt;character&gt;
## bonf       stats         3.6.0           FALSE        pval "bonferroni"
## BH         stats         3.6.0           FALSE        pval         "BH"
##         meta.type session.idx
##       &lt;character&gt;   &lt;numeric&gt;
## bonf FWER control           1
## BH    FDR control           1</code></pre>
<p>While all methods in this example had the <code>meta =</code> option specified, this is not necessary. It is completely acceptable to specify the <code>meta =</code> parameter for only a subset of methods. Missing entries will be recorded as <code>NA</code>s in the <code>colData()</code>.</p>
</div>
<div id="specifying-version-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#specifying-version-metadata" class="anchor"></a>Specifying Version Metadata</h2>
<p>Arguably, two of the most important pieces of metadata stored in the <code>colData()</code> of the <em>SummarizedBenchmark</em> returned by <code><a href="../reference/buildBench.html">buildBench()</a></code> are the relevant package name and version (<code>pkg_name</code>, <code>pkg_vers</code>). Determining the package name and version requires the primary “workhorse” function of the method be directly specified as <code>func =</code> in the <code><a href="../reference/addMethod.html">addMethod()</a></code> call. In some cases, this may not be possible, e.g. if the “workhorse” function is a wrapper. However, there still might exist an important function for which we would like to track the package name and version. The <code>meta</code> parameter can help.</p>
<p>The <code>meta =</code> parameter will handle the following named list entries as special values: <code>pkg_name</code>, <code>pkg_vers</code>, <code>pkg_func</code>. First, if values are specified for <code>pkg_name</code> and <code>pkg_vers</code> in <code>meta =</code>, these will overwrite the values determined from <code>func =</code>. To trace the source of <code>pkg_name</code> and <code>pkg_vers</code> information, the <code>func.pkg.manual</code> column of the <code>colData</code> will be set to <code>TRUE</code> (the default value is <code>FALSE</code>). Alternatively, a function can be passed to <code>meta =</code> as <code>pkg_func</code>. This function will be used to determine both <code>pkg_name</code> and <code>pkg_vers</code>, and will take precendence over manually specified <code>pkg_name</code> and <code>pkg_vers</code> values. If <code>pkg_func</code> is specified, it will be included in the <code>colData()</code> as a new column with the same name, and the <code>vers_src</code> column will be set to <code>"meta_func"</code>. **Note: the function should be wrapped in <code><a href="https://rlang.r-lib.org/reference/quotation.html">rlang::quo</a></code> to be properly parsed.</p>
<p>We illustrate the behavior when using either <code>pkg_func</code> or <code>pkg_name</code> and <code>pkg_vers</code> with the <code>meta</code> optional parameter using the same multiple testing example from above. First, notice that Storey’s q-value method was previously included in the <em>BenchDesign</em> with a <code>post =</code> argument to extract q-values from the result of calling <code><a href="https://www.rdocumentation.org/packages/qvalue/topics/qvalue">qvalue::qvalue</a></code> on the data.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="kw"><a href="../reference/printMethod.html">printMethod</a></span>(b, <span class="st">"qv"</span>)</a></code></pre></div>
<pre><code>## qv --------------------------------------------------------- 
## BenchDesign Method (BDMethod) ------------------------------ 
##   method: qvalue::qvalue
##     function (p, fdr.level = NULL, pfdr = FALSE, lfdr.out... 
##   parameters:
##     p : pval 
##   post:
##     default : function (x) { ... 
##   meta:
##     none</code></pre>
<p>We can avoid specifying <code>post =</code> by using the following wrapper function in place of the direct call to <code><a href="https://www.rdocumentation.org/packages/qvalue/topics/qvalue">qvalue::qvalue</a></code>.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1">qv_wrapper &lt;-<span class="st"> </span><span class="cf">function</span>(...) {</a>
<a class="sourceLine" id="cb46-2" data-line-number="2">    qvalue<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/qvalue/topics/qvalue">qvalue</a></span>(...)<span class="op">$</span>qvalues</a>
<a class="sourceLine" id="cb46-3" data-line-number="3">}</a>
<a class="sourceLine" id="cb46-4" data-line-number="4"></a>
<a class="sourceLine" id="cb46-5" data-line-number="5">b_qv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>(<span class="dt">data =</span> tdat) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb46-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"qv"</span>,</a>
<a class="sourceLine" id="cb46-7" data-line-number="7">              <span class="dt">func =</span> qvalue<span class="op">::</span>qvalue,</a>
<a class="sourceLine" id="cb46-8" data-line-number="8">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval),</a>
<a class="sourceLine" id="cb46-9" data-line-number="9">              <span class="dt">post =</span> <span class="cf">function</span>(x) { x<span class="op">$</span>qvalues }) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb46-10" data-line-number="10"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"qv_wrapper"</span>,</a>
<a class="sourceLine" id="cb46-11" data-line-number="11">              <span class="dt">func =</span> qv_wrapper,</a>
<a class="sourceLine" id="cb46-12" data-line-number="12">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval))</a></code></pre></div>
<p>However, we see that while the results for the two methods are the same, the <code>colData()</code> of the resulting <em>SummarizedBenchmark</em> object is missing the valuable package name and version information for the wrapper-based approach.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1">sb_qv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(b_qv)</a>
<a class="sourceLine" id="cb47-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="kw">assay</span>(sb_qv))</a></code></pre></div>
<pre><code>##                qv   qv_wrapper
## [1,] 0.0079813228 0.0079813228
## [2,] 0.3290660527 0.3290660527
## [3,] 0.0003753518 0.0003753518
## [4,] 0.0442094809 0.0442094809
## [5,] 0.1716134119 0.1716134119
## [6,] 0.1757029652 0.1757029652</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" data-line-number="1"><span class="kw">colData</span>(sb_qv)</a></code></pre></div>
<pre><code>## DataFrame with 2 rows and 5 columns
##               func.pkg func.pkg.vers func.pkg.manual     param.p
##            &lt;character&gt;   &lt;character&gt;       &lt;logical&gt; &lt;character&gt;
## qv              qvalue        2.17.0           FALSE        pval
## qv_wrapper          NA            NA           FALSE        pval
##            session.idx
##              &lt;numeric&gt;
## qv                   1
## qv_wrapper           1</code></pre>
<p>This can be fixed by specifying either the <code>pkg_func</code> or <code>pkg_name</code> and <code>pkg_vers</code> metadata values.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1">b_qv &lt;-<span class="st"> </span>b_qv <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"qv_pkgfunc"</span>,</a>
<a class="sourceLine" id="cb51-3" data-line-number="3">              <span class="dt">func =</span> qv_wrapper,</a>
<a class="sourceLine" id="cb51-4" data-line-number="4">              <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">pkg_func =</span> qvalue<span class="op">::</span>qvalue),</a>
<a class="sourceLine" id="cb51-5" data-line-number="5">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-6" data-line-number="6"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"qv_pkgname"</span>,</a>
<a class="sourceLine" id="cb51-7" data-line-number="7">              <span class="dt">func =</span> qv_wrapper,</a>
<a class="sourceLine" id="cb51-8" data-line-number="8">              <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">pkg_name =</span> <span class="st">"qvalue"</span>, <span class="dt">pkg_vers =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/packageDescription">packageVersion</a></span>(<span class="st">"qvalue"</span>))),</a>
<a class="sourceLine" id="cb51-9" data-line-number="9">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">p =</span> pval))</a>
<a class="sourceLine" id="cb51-10" data-line-number="10"></a>
<a class="sourceLine" id="cb51-11" data-line-number="11">sb_qv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(b_qv)</a>
<a class="sourceLine" id="cb51-12" data-line-number="12"><span class="kw">colData</span>(sb_qv)</a></code></pre></div>
<pre><code>## DataFrame with 4 rows and 5 columns
##               func.pkg func.pkg.vers func.pkg.manual     param.p
##            &lt;character&gt;   &lt;character&gt;       &lt;logical&gt; &lt;character&gt;
## qv              qvalue        2.17.0           FALSE        pval
## qv_wrapper          NA            NA           FALSE        pval
## qv_pkgfunc      qvalue        2.17.0            TRUE        pval
## qv_pkgname      qvalue        2.17.0            TRUE        pval
##            session.idx
##              &lt;numeric&gt;
## qv                   1
## qv_wrapper           1
## qv_pkgfunc           1
## qv_pkgname           1</code></pre>
<p>The <code>func.pkg</code> and <code>func.pkg.vers</code> columns of the <code>colData()</code> are now populated with the correct information, and furthermore, the <code>func.pkg.manual</code> column is set to <code>TRUE</code> for the two latter methods.</p>
</div>
<div id="modifying-methods-in-a-benchdesign" class="section level2">
<h2 class="hasAnchor">
<a href="#modifying-methods-in-a-benchdesign" class="anchor"></a>Modifying Methods in a BenchDesign</h2>
<p>Modifying the defintion of a method after it has been added to a <em>BenchDesign</em> is supported by the <code><a href="../reference/modifyMethod.html">modifyMethod()</a></code> function. The <em>BenchDesign</em> object created in the multiple testing example above, <code>b</code>, includes a method called <code>qv</code>.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" data-line-number="1"><span class="kw"><a href="../reference/printMethod.html">printMethod</a></span>(b, <span class="st">"qv"</span>)</a></code></pre></div>
<pre><code>## qv --------------------------------------------------------- 
## BenchDesign Method (BDMethod) ------------------------------ 
##   method: qvalue::qvalue
##     function (p, fdr.level = NULL, pfdr = FALSE, lfdr.out... 
##   parameters:
##     p : pval 
##   post:
##     default : function (x) { ... 
##   meta:
##     none</code></pre>
<p>Suppose we wish to specify an additional <code>pi0 =</code> parameter for the <code><a href="https://www.rdocumentation.org/packages/qvalue/topics/qvalue">qvalue::qvalue</a></code> method as well as add a new metadata tag. This can be accomplished by passing both new parameters to the <code>params =</code> parameter of <code><a href="../reference/modifyMethod.html">modifyMethod()</a></code> as a <code><a href="https://rlang.r-lib.org/reference/quotation.html">rlang::quos(..)</a></code> list. By default, all entries in the <code><a href="https://rlang.r-lib.org/reference/quotation.html">rlang::quos(..)</a></code> list are assumed to be parameters for the main function of the method. However, <code>func</code>, <code>post</code>, or <code>meta</code> values can also be modified by included in their new values in the list using the special keywords, <code>bd.func =</code>, <code>bd.post =</code>, or <code>bd.meta =</code>. In the example below, we both add a new parameter , <code>pi0 =</code> and modify the <code>meta =</code> specification.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" data-line-number="1">new_params &lt;-<span class="st"> </span>rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">pi0 =</span> <span class="fl">0.9</span>, <span class="dt">bd.meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">type =</span> <span class="st">"FDR control"</span>))</a>
<a class="sourceLine" id="cb55-2" data-line-number="2">b_modqv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/modifyMethod.html">modifyMethod</a></span>(b, <span class="dt">label =</span> <span class="st">"qv"</span>, <span class="dt">params =</span> new_params)</a>
<a class="sourceLine" id="cb55-3" data-line-number="3"></a>
<a class="sourceLine" id="cb55-4" data-line-number="4"><span class="kw"><a href="../reference/printMethod.html">printMethod</a></span>(b_modqv, <span class="st">"qv"</span>)</a></code></pre></div>
<pre><code>## qv --------------------------------------------------------- 
## BenchDesign Method (BDMethod) ------------------------------ 
##   method: qvalue::qvalue
##     function (p, fdr.level = NULL, pfdr = FALSE, lfdr.out... 
##   parameters:
##     p : pval 
##     pi0 : 0.9 
##   post:
##     default : function (x) { ... 
##   meta:
##     type : "FDR control"</code></pre>
<p>We can see that the original method parameters, namely <code>p = pval</code>, is still defined for method, while the new method parameter <code>pi0 = 0.9</code> has been added along with the new metadata specification. However, sometimes it may be desirable to completely overwrite all function parameters for a method, e.g. if we decide to completely redefine the parameters of a method. This may occur if some parameters were optional and originally specified, but no longer necessary. All function parameters can be overwritten by specifying <code>.overwrite = TRUE</code>.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" data-line-number="1">b_modqv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/modifyMethod.html">modifyMethod</a></span>(b, <span class="dt">label =</span> <span class="st">"qv"</span>, <span class="dt">params =</span> new_params, <span class="dt">.overwrite =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb57-2" data-line-number="2"></a>
<a class="sourceLine" id="cb57-3" data-line-number="3"><span class="kw"><a href="../reference/printMethod.html">printMethod</a></span>(b_modqv, <span class="st">"qv"</span>)</a></code></pre></div>
<pre><code>## qv --------------------------------------------------------- 
## BenchDesign Method (BDMethod) ------------------------------ 
##   method: qvalue::qvalue
##     function (p, fdr.level = NULL, pfdr = FALSE, lfdr.out... 
##   parameters:
##     pi0 : 0.9 
##   post:
##     default : function (x) { ... 
##   meta:
##     type : "FDR control"</code></pre>
<p>Now the originally specified <code>p = pval</code> is no longer included in the method definition. Obviously, this is not reasonable in this setting as the <code>p</code> parameter is a necessary parameter to run the method. Note that specifications of <code>bd.meta =</code> will always overwrite the existing metadata values of the method, even when <code>.overwrite = FALSE</code>.</p>
</div>
<div id="duplicating-methods-in-a-benchdesign" class="section level2">
<h2 class="hasAnchor">
<a href="#duplicating-methods-in-a-benchdesign" class="anchor"></a>Duplicating Methods in a BenchDesign</h2>
<p>In addition to comparing multiple methods, a benchmark study may also involve comparing a single method across several parameter settings. The <code><a href="../reference/expandMethod.html">expandMethod()</a></code> function provides the capability to take a method already defined in the <em>BenchDesign</em>, and <em>expand</em> it to multiple methods with differing parameter values in the <em>BenchDesign</em> object. In the following example, <code><a href="../reference/expandMethod.html">expandMethod()</a></code> is used to duplicate the <code>qv</code> method with only the <code>pi0</code> parameter modified. By default, the expanded method, in this case <code>qv</code>, is dropped. To prevent this from happening, <code>.replace = FALSE</code> can be specified.</p>
<p>Expanding the <code>qv</code> method with <code><a href="../reference/expandMethod.html">expandMethod()</a></code> is accomplished by specifying any new parameters of the variant methods to <code>params =</code> as a <code><a href="https://rlang.r-lib.org/reference/quotation.html">rlang::quos(..)</a></code> list. Here, two new variants of the <code>qv</code> method, <code>qv-p90</code> and <code>qv-p10</code>, are created with new values of the <code>pi0</code> parameter.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1">b_expand &lt;-<span class="st"> </span><span class="kw"><a href="../reference/expandMethod.html">expandMethod</a></span>(b, <span class="dt">label =</span> <span class="st">"qv"</span>, </a>
<a class="sourceLine" id="cb59-2" data-line-number="2">                         <span class="dt">params =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">qv_p90 =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">pi0 =</span> <span class="fl">0.90</span>),</a>
<a class="sourceLine" id="cb59-3" data-line-number="3">                                       <span class="dt">qv_p10 =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">pi0 =</span> <span class="fl">0.10</span>)),</a>
<a class="sourceLine" id="cb59-4" data-line-number="4">                         <span class="dt">.replace =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb59-5" data-line-number="5"></a>
<a class="sourceLine" id="cb59-6" data-line-number="6">b_expand</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     type: md5hash 
##     MD5 hash: 09867f2696d3178a8f3bc038917042e1 
##   benchmark methods:
##     method:      BH; func: p.adjust       
##     method:      qv; func: qvalue::qvalue 
##     method:    bonf; func: p.adjust       
##     method:  qv_p90; func: qvalue::qvalue 
##     method:  qv_p10; func: qvalue::qvalue</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" data-line-number="1"><span class="kw"><a href="../reference/printMethod.html">printMethod</a></span>(b_expand, <span class="st">"qv_p90"</span>)</a></code></pre></div>
<pre><code>## qv_p90 ----------------------------------------------------- 
## BenchDesign Method (BDMethod) ------------------------------ 
##   method: qvalue::qvalue
##     function (p, fdr.level = NULL, pfdr = FALSE, lfdr.out... 
##   parameters:
##     p : pval 
##     pi0 : 0.9 
##   post:
##     default : function (x) { ... 
##   meta:
##     none</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" data-line-number="1"><span class="kw"><a href="../reference/printMethod.html">printMethod</a></span>(b_expand, <span class="st">"qv_p10"</span>)</a></code></pre></div>
<pre><code>## qv_p10 ----------------------------------------------------- 
## BenchDesign Method (BDMethod) ------------------------------ 
##   method: qvalue::qvalue
##     function (p, fdr.level = NULL, pfdr = FALSE, lfdr.out... 
##   parameters:
##     p : pval 
##     pi0 : 0.1 
##   post:
##     default : function (x) { ... 
##   meta:
##     none</code></pre>
<p>When only a single parameter is modifed as in the example above, an alternative shorthand is provided. The parameter can be specified with <code>onlyone =</code> and the corresponding values specified with the new method names to <code>params =</code>.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" data-line-number="1">b_expand &lt;-<span class="st"> </span><span class="kw"><a href="../reference/expandMethod.html">expandMethod</a></span>(b, <span class="dt">label =</span> <span class="st">"qv"</span>, <span class="dt">onlyone =</span> <span class="st">"pi0"</span>,</a>
<a class="sourceLine" id="cb65-2" data-line-number="2">                         <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">qv_p90 =</span> <span class="fl">0.90</span>,</a>
<a class="sourceLine" id="cb65-3" data-line-number="3">                                              <span class="dt">qv_p10 =</span> <span class="fl">0.10</span>),</a>
<a class="sourceLine" id="cb65-4" data-line-number="4">                         <span class="dt">.replace =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb65-5" data-line-number="5"></a>
<a class="sourceLine" id="cb65-6" data-line-number="6">b_expand</a></code></pre></div>
<pre><code>## BenchDesign ------------------------------------------------ 
##   benchmark data:
##     type: md5hash 
##     MD5 hash: 09867f2696d3178a8f3bc038917042e1 
##   benchmark methods:
##     method:      BH; func: p.adjust       
##     method:      qv; func: qvalue::qvalue 
##     method:    bonf; func: p.adjust       
##     method:  qv_p90; func: qvalue::qvalue 
##     method:  qv_p10; func: qvalue::qvalue</code></pre>
<p>The resulting modified <em>BenchDesign</em> object is the same.</p>
</div>
</div>
<div id="summarizedbenchmark-features" class="section level1">
<h1 class="hasAnchor">
<a href="#summarizedbenchmark-features" class="anchor"></a>SummarizedBenchmark Features</h1>
<p>Aside from the standard <em>SummarizedBenchmark</em></p>
<div id="manually-constructing" class="section level2">
<h2 class="hasAnchor">
<a href="#manually-constructing" class="anchor"></a>Manually Constructing</h2>
<p>Throughout our documentation, <em>SummarizedBenchmark</em> objects are constructed exclusively using our recommended workflow by running <code><a href="../reference/buildBench.html">buildBench()</a></code> with a <em>BenchDesign</em> object. This recommended approach enables users to perform benchmark comparisons while automatically keeping track of parameters and software metadata. However, it is also possible to skip this recommended approach and construct a <em>SummarizedBenchmark</em> object directly from standard <code>S3</code> data objects.</p>
<p>Using data from the <em><a href="https://bioconductor.org/packages/3.10/iCOBRA">iCOBRA</a></em>package <span class="citation">(Soneson and Robinson 2016)</span>, this part of the vignette demonstrates how to build <em>SummarizedBenchmark</em> objects from <code>S3</code> objects. The dataset contains differential expression results of three different methods (<em><a href="https://bioconductor.org/packages/3.10/limma">limma</a></em>, <em><a href="https://bioconductor.org/packages/3.10/edgeR">edgeR</a></em> and <em><a href="https://bioconductor.org/packages/3.10/DESeq2">DESeq2</a></em>) applied to a simulated RNA-seq dataset.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(iCOBRA)</a>
<a class="sourceLine" id="cb67-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(cobradata_example)</a></code></pre></div>
<p>The process of building a <em>SummarizedBenchmark</em> object is similar to the process of constructing a <em>SummarizedExperiment</em> object. To build a <em>SummarizedBenchmark</em> object, three main objects are required: (1) a list where each element corresponds to a <em>data.frame</em>, (2) a <em>DataFrame</em> with annotations of the methods, and (3) when available, a <em>DataFrame</em> of ground truths.</p>
<p>In the <em>SummarizedBenchmark</em> object, each output of the methods is considered a different <code>assay</code>. For example, using the differential expression dataset example, we can define two assays, q-values and estimated log fold changes. For each <code>assay</code>, we arrange the output of the different methods as a matrix where each column corresponds to a method and each row corresponds to each feature (in this case, genes). We will need a list in which each of it’s element corresponds to an assay.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb68-1" data-line-number="1">assays &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">qvalue =</span> cobradata_example<span class="op">@</span>padj,</a>
<a class="sourceLine" id="cb68-2" data-line-number="2">               <span class="dt">logFC =</span> cobradata_example<span class="op">@</span>score)</a>
<a class="sourceLine" id="cb68-3" data-line-number="3">assays[[<span class="st">"qvalue"</span>]]<span class="op">$</span>DESeq2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/p.adjust">p.adjust</a></span>(cobradata_example<span class="op">@</span>pval<span class="op">$</span>DESeq2, <span class="dt">method =</span> <span class="st">"BH"</span>)</a>
<a class="sourceLine" id="cb68-4" data-line-number="4"></a>
<a class="sourceLine" id="cb68-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(assays[[<span class="st">"qvalue"</span>]], <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##                     edgeR      voom     DESeq2
## ENSG00000000457 0.2833295 0.1375146 0.04760152
## ENSG00000000460 0.2700376 0.2826398 0.08717737
## ENSG00000000938 0.9232794 0.9399669 0.99856748</code></pre>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb70-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(assays[[<span class="st">"logFC"</span>]], <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##                      edgeR       voom      DESeq2
## ENSG00000000457 0.22706080 0.23980596 0.274811698
## ENSG00000000460 0.31050478 0.27677537 0.359226203
## ENSG00000000938 0.04985025 0.03344113 0.001526355</code></pre>
<p>Since these are simulated data, the ground truths for both assays are known and available. We can format these as a <em>DataFrame</em> where each column corresponds to an assay and each row corresponds to a feature.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb72-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(S4Vectors)</a>
<a class="sourceLine" id="cb72-2" data-line-number="2"></a>
<a class="sourceLine" id="cb72-3" data-line-number="3">groundTruth &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/DataFrame-class">DataFrame</a></span>(cobradata_example<span class="op">@</span>truth[, <span class="kw"><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class">c</a></span>(<span class="st">"status"</span>, <span class="st">"logFC"</span>)])</a>
<a class="sourceLine" id="cb72-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(groundTruth) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(assays)</a>
<a class="sourceLine" id="cb72-5" data-line-number="5">groundTruth &lt;-<span class="st"> </span>groundTruth[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(assays[[<span class="dv">1</span>]]), ]</a>
<a class="sourceLine" id="cb72-6" data-line-number="6"></a>
<a class="sourceLine" id="cb72-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class">head</a></span>(groundTruth)</a></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##                    qvalue              logFC
##                 &lt;integer&gt;          &lt;numeric&gt;
## ENSG00000000457         0                  0
## ENSG00000000460         0                  0
## ENSG00000000938         0                  0
## ENSG00000000971         0                  0
## ENSG00000001460         1   3.93942204325368
## ENSG00000001461         1 0.0602244207363294</code></pre>
<p>Finally, the method names are also reformatted as a <em>DataFrame</em> of simple annotation data of the methods.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb74-1" data-line-number="1">colData &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/DataFrame-class">DataFrame</a></span>(<span class="dt">method =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(assays[[<span class="dv">1</span>]]))</a>
<a class="sourceLine" id="cb74-2" data-line-number="2">colData</a></code></pre></div>
<pre><code>## DataFrame with 3 rows and 1 column
##        method
##   &lt;character&gt;
## 1       edgeR
## 2        voom
## 3      DESeq2</code></pre>
<p>A <em>SummarizedBenchmark</em> is built using the following call to the <code><a href="../reference/SummarizedBenchmark.html">SummarizedBenchmark()</a></code> constructor.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb76-1" data-line-number="1">sb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SummarizedBenchmark.html">SummarizedBenchmark</a></span>(<span class="dt">assays =</span> assays, <span class="dt">colData =</span> colData, <span class="dt">groundTruth =</span> groundTruth)</a></code></pre></div>
</div>
</div>
<div id="next-steps" class="section level1">
<h1 class="hasAnchor">
<a href="#next-steps" class="anchor"></a>Next Steps</h1>
<p>This vignette described the structure of the <em>BenchDesign</em> and <em>SummarizedBenchmark</em> classes which underly the <code>SummarizedBenchmark</code> package, as well as several complex and non-standard features of both classes. Examples of using these classes and the <code>SummarizedBenchmark</code> framework can be found in the <strong>SummarizedBenchmark: Full Case Study</strong> vignette as well as in the other <strong>Case Study</strong> vignettes. More advanced features are also described in various <strong>Feature</strong> vignettes.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Soneson_2016">
<p>Soneson, Charlotte, and Mark D Robinson. 2016. “iCOBRA: Open, Reproducible, Standardized and Live Method Benchmarking.” <em>Nature Methods</em> 13 (4):283–83. <a href="https://doi.org/10.1038/nmeth.3805" class="uri">https://doi.org/10.1038/nmeth.3805</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#benchdesign">BenchDesign</a></li>
      <li><a href="#summarizedbenchmark">SummarizedBenchmark</a></li>
      <li><a href="#reproducibility-and-replicability">Reproducibility and Replicability</a></li>
      <li><a href="#benchdesign-features">BenchDesign Features</a></li>
      <li><a href="#summarizedbenchmark-features">SummarizedBenchmark Features</a></li>
      <li><a href="#next-steps">Next Steps</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://alejandroreyes.org">Alejandro Reyes</a>, <a href="https://www.pkimes.com">Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
